src/huggingface_hub/snapshot_download.py

```diff
@@ -53,18 +64,86 @@ def snapshot_download(
 
-    storage_folder = os.path.join(
-        cache_dir, repo_id.replace("/", REPO_ID_SEPARATOR) + "." + model_info.sha
-    )
+        # 3) cached repos of format <repo_id>.<any-branch>.{revision}
+        # -> in this case {revision} also has to be a commit sha
+        repo_folders_branch_commit = glob(repo_folders_prefix + ".*." + revision)
+
+        # combine all possible fetched cached repos
+        repo_folders = (
+            repo_folders_branch + repo_folders_commit_only + repo_folders_branch_commit
+        )
+
+        if len(repo_folders) == 0:
+            raise ValueError(
+                "Cannot find the requested files in the cached path and outgoing traffic has been"
+                " disabled. To enable model look-ups and downloads online, set 'local_files_only'"
+                " to False."
+            )
 
-    for model_file in model_info.siblings:
-        url = hf_hub_url(
-            repo_id, filename=model_file.rfilename, revision=model_info.sha
+        # check if repo id was previously cached from a commit sha revision
+        # and passed {revision} is not a commit sha
+        # in this case snapshotting repos locally might lead to unexpected
+        # behavior the user should be warned about
+        repo_folders_commit_any = set(glob(repo_folders_prefix + ".*")) - set(
+            glob(repo_folders_prefix + ".*.*")
         )
-        relative_filepath = os.path.join(*model_file.rfilename.split("/"))
+        if (
+            len(repo_folders_commit_any) > 0
+            and len(repo_folders_commit_only + repo_folders_branch_commit) == 0
+        ):
+            logger.warn(
+                f"The repo {repo_id} was previously downloaded from a commit hash revision "
+                f"and has created the following cached directories {repo_folders_commit_any}."
+                f" In this case, trying to load a repo from the branch {revision} in offline "
+                "mode might lead to unexpected behavior by not taking into account the latest "
+                "commits."
+            )
+
+        # find last modified folder
+        storage_folder = max(repo_folders, key=os.path.getmtime)
+
+        # get commit sha
+        repo_id_sha = storage_folder.split(".")[-1]
+        model_files = os.listdir(storage_folder)
+    else:
+        # if we have internet connection we retrieve the correct folder name from the huggingface api
+        _api = HfApi()
+        model_info = _api.model_info(repo_id=repo_id, revision=revision, token=token)
+
+        storage_folder = os.path.join(cache_dir, repo_id_flattened + "." + revision)
+
+        # if passed revision is not identical to the commit sha
+        # then revision has to be a branch name, e.g. "main"
+        # in this case make sure that the branch name is included
+        # cached storage folder name
+        if revision != model_info.sha:
+            storage_folder += f".{model_info.sha}"
```

Comment:
I wonder if this is necessary or superfluous - does it hurt to keep things simple at the cost of having a duplicate of sha in the saved folder name?

Reply:
After f2f explanation I understand why this is necessary :)